cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray back() called on empty async_mdarray!");
+    return begin()[size() - 1];
+  }
+
+  //! @brief Returns a reference to the last element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_reference back() const noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray back() called on empty async_mdarray!");
+    return begin()[size() - 1];
+  }
+  //! @}
+
+  //! @addtogroup capacity
+  //! @{
+  //! @brief Returns the current number of elements stored in the async_mdarray.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI size_type size() const noexcept
+  {
+    return __size(__mapping_.extents());
+  }
+
+  //! @brief Returns true if the async_mdarray is empty.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI bool empty() const noexcept
+  {
+    return size() == 0;
+  }
+
+  //! @brief Returns the capacity of the current allocation of the async_mdarray..
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI size_type capacity() const noexcept
+  {
+    return size();
+  }
+
+  //! @brief Returns the maximal size of the async_mdarray.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI constexpr size_type max_size() const noexcept
+  {
+    return static_cast<size_type>((_CUDA_VSTD::numeric_limits<difference_type>::max)());
+  }
+
+  //! @rst
+  //! Returns a \c const reference to the :ref:`any_resource <cudax-memory-resource-any-resource>`
+  //! that holds the memory resource used to allocate the async_mdarray
+  //! @endrst
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const __resource_t& get_memory_resource() const noexcept
+  {
+    return __buf_.get_memory_resource();
+  }
+
+  //! @brief Returns the stored stream
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI constexpr ::cuda::stream_ref get_stream() const noexcept
+  {
+    return __buf_.get_stream();
+  }
+
+  //! @brief Replaces the stored stream
+  //! @param __new_stream the new stream
+  //! @note Always synchronizes with the old stream
+  _CCCL_HIDE_FROM_ABI constexpr void change_stream(::cuda::stream_ref __new_stream)
+  {
+    __buf_.change_stream(__new_stream);
+  }
+
+  //! @brief Replaces the stored stream
+  //! @param __new_stream the new stream
+  //! @warning This does not synchronize between \p __new_stream and the current stream. It is the user's responsibility
+  //! to ensure proper stream order going forward
+  _CCCL_HIDE_FROM_ABI constexpr void change_stream_unsynchronized(::cuda::stream_ref __new_stream) noexcept
+  {
+    __buf_.change_stream_unsynchronized(__new_stream);
+  }
+
+  //! @brief Returns the execution policy
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI constexpr __policy_t get_execution_policy() const noexcept
+  {
+    return __policy_;
+  }
+
+  //! @brief Replaces the currently used execution policy
+  //! @param __new_policy the new policy
+  _CCCL_HIDE_FROM_ABI constexpr void set_execution_policy(__policy_t __new_policy) noexcept
+  {
+    __policy_ = __new_policy;
+  }
+
```

Comment:
Changing the execution policy or stream after construction feels like changing the allocator after construction -- it's not something that one can do to `vector`, and it makes reasoning about code that takes `async_mdarray`& difficult.

Should we consider an alternate design, that makes it easier to move from this async_mdarray into a new async_mdarray with a different execution policy?

```suggestion
```

Reply:
I do not follow you here. 

The execution policy determines where future operations on the mdarray happen. In what regard would that be problematic?

Reply:
I wrote this above; I'll just repeat it here for clarity.

1. (Re)allocation and future work are coupled through copy assignment.
2. `std::vector` has the invariant that it always has the same allocator throughout its lifetime.
3. If (re)allocation and future work are coupled, then should we also consider adding the invariant that a container always has the same allocator and stream throughout its lifetime?
4. Move construction is relatively cheap and can preserve existing allocations and values.
5. Therefore, should we consider a design in which users can "change" the allocator and/or stream by move-constructing a new container with the desired allocator and/or stream?  This would change the allocator and/or stream for future operations; the current allocation wouldn't necessarily change.

Here's an example of (5).

```c++
env e{allocator, stream}; // whatever the syntax may be
async_mdarray<float, std::dims<2, int>> A{env, num_rows, num_cols};
// ... code ...
async_mdarray<float, std::dims<2, int>> B = /* from somewhere */;
// ... more code ...
env e2{new_allocator, new_stream}; // either or both may change
async_mdarray<float, std::dims<2, int>> A2{e2, std::move(A)}; 
// Assignment from B uses A2's new allocator and new stream.
A2 = B;
```

Reply:
>  This would change the allocator and/or stream for future operations; the current allocation wouldn't necessarily change.

This would be like changing the scheduler in std::execution: it affects things that happen after the change.  I find this to be a reasonable interpretation of the name "`async_mdarray`."

Reply:
> The execution policy determines where future operations on the mdarray happen. In what regard would that be problematic?

One feature of `std::vector` is that I can't change its allocator after construction.  In `async_mdarray`, copy assignment couples (re)allocation and asynchronous execution (copying or fill).  In a hypothetical `async_vector`, both copy assignment and resizing would couple allocation and asynchronous execution.  (Re)allocation also might depend on the stream, so users might reasonably expect that if they could change the stream, they could also change the allocator.